synctube.go

```diff
@@ -140,13 +140,53 @@ func (r *Replication) shutdownEventChannel() {
 
 func (r Replication) targetCheckpointAddress() string {
 
-	// TODO: this needs to take into account other aspects
-	// of replication (filters, filterparams, etc)
+	// Calculate a target checkpoint address / local document ID for storing checkpoint "local doc" on the remote
+	// target Sync Gateway.  This checkpoint address / local document ID should be as unique as possible, to avoid
+	// issues like #16 where different replications erroneously use the same checkpoint and docs
+	// fail to transfer.  This code tries to emulate the behavior in Couchbase Lite iOS
+	// CBL_ReplicatorSettings.m::remoteCheckpointDocIDForLocalUUID (http://bit.ly/2a3HMdx) as much as possible.
+
+	// serialize the parameters to JSON
+	replicationParamsJsonBytes, err := json.Marshal(r.Parameters)
+	if err != nil {
+		// TODO: this should return an error rather than panicking
+		clog.Panic("Unable to generate checkpoint address. Err: %v", err)
+	}
+	replicationParamsJson := string(replicationParamsJsonBytes)
+
+	// We don't have a local UUID for the local database, because there is no local database
+	// in the sg-replicate case.  However, we should avoid the case where different machines
+	// are running sg-replicate with the exact same parameters, and then step on each other's
+	// remote checkpoints.  So to try to guard against that, try to append the hostname to the string
+	// that will be used to calculate the hash.  Of course, this isn't perfect, especially
+	// when this is running on networks where the System Administrator sets the hostname of
+	// all machines to 'localhost'.
+	hostname, err := os.Hostname()
+	if err == nil {
+		replicationParamsJson = fmt.Sprintf("%v%v", replicationParamsJson, hostname)
+	}
+
+	// The pattern for generating a hash is `sha1.New()`,
+	// `sha1.Write(bytes)`, then `sha1.Sum([]byte{})`.
+	// Here we start with a new hash.
+	h := sha1.New()
+
+	// `Write` expects bytes. If you have a string `s`,
+	// use `[]byte(s)` to coerce it to bytes.
+	h.Write([]byte(replicationParamsJson))
+
+	// This gets the finalized hash result as a byte
+	// slice. The argument to `Sum` can be used to append
+	// to an existing byte slice: it usually isn't needed.
+	bs := h.Sum(nil)
+
```

Comment:
For this use case I thought we could just use sha1.Sum([]byte(replicationParamsJson), and not need the New/Write/Sum.  e.g. https://golang.org/pkg/crypto/sha1/#Sum